lib/src/graph.dart

```diff
@@ -270,27 +270,150 @@ class Graph {
-    Set objs = {};
+  ///
+  /// If `sub` is provided, it checks for the subject.
+  /// If `pre` is provided, it checks for the predicate.
+  /// If both are provided, it checks for both subject and predicate.
+  /// If neither is provided, it returns all objects in the triples.
+  Set<dynamic> objects({URIRef? sub, URIRef? pre}) {
+    // Initialize an empty set to store the objects.
+    Set<dynamic> objs = {};
+
+    // Iterate over all triples in the graph.
     for (Triple t in triples) {
-      if (t.sub == sub && t.pre == pre) {
+      // Check conditions based on the presence of sub and pre.
+      if (sub != null && pre != null) {
+        // Both sub and pre are provided.
+        if (t.sub == sub && t.pre == pre) {
+          objs.add(t.obj);
+        }
+      } else if (sub != null) {
+        // Only sub is provided.
+        if (t.sub == sub) {
+          objs.add(t.obj);
+        }
+      } else if (pre != null) {
+        // Only pre is provided.
+        if (t.pre == pre) {
+          objs.add(t.obj);
+        }
+      } else {
+        // Neither sub nor pre is provided, return all objects.
         objs.add(t.obj);
       }
     }
+
+    // Return the set of objects.
     return objs;
   }
 
+  /// Finds all predicates which have a certain subject and object.
+  ///
+  /// If `sub` is provided, it checks for the subject.
+  /// If `obj` is provided, it checks for the object.
+  /// If both are provided, it checks for both subject and object.
+  /// If neither is provided, it returns all predicates in the triples.
+  Set<URIRef> predicates({URIRef? sub, dynamic obj}) {
+    // Initialize an empty set to store the predicates.
+    Set<URIRef> pres = {};
+
+    // Iterate over all triples in the graph.
+    for (Triple t in triples) {
+      // Check conditions based on the presence of sub and obj.
+      if (sub != null && obj != null) {
+        // Both sub and obj are provided.
+        if (t.sub == sub && t.obj == obj) {
+          pres.add(t.pre);
+        }
+      } else if (sub != null) {
+        // Only sub is provided.
+        if (t.sub == sub) {
+          pres.add(t.pre);
+        }
+      } else if (obj != null) {
+        // Only obj is provided.
+        if (t.obj == obj) {
+          pres.add(t.pre);
+        }
+      } else {
+        // Neither sub nor obj is provided, return all predicates.
+        pres.add(t.pre);
+      }
+    }
+
+    // Return the set of predicates.
+    return pres;
+  }
+
+  /// Finds all triples which have a certain value.
+  ///
+  /// This method checks if the given [value] matches any of the components
+  /// (subject, predicate, or object) of the triples in the graph. It returns
+  /// a set of triples where any of these components matches the [value].
+  ///
+  /// Example usage:
+  /// ```dart
+  /// final value = 'exampleValue';
+  /// final matchingTriples = values(value);
+  /// print(matchingTriples);
+  /// ```
+  Set<Triple> tripleValues(String value) {
```

Comment:
Perhaps naming it `triples()` instead of `tripleValues` if applicable.

Reply:
Hi @cdawei, the name 'triples' is already defined in the `parser/grammar_parser.dart`. The name can be renamed to 'matchTriples()'.